File: bridge.go
package main

import "fmt"

// Interfaces
type Shape interface {
    Draw()
}

type Renderer interface {
    RenderCircle()
    RenderSquare()
}

// Bridge between shape and renderer
type AbstractShapeRenderer struct {
    renderer Renderer
}

func (asr *AbstractShapeRenderer) SetRenderer(renderer Renderer) {
    asr.renderer = renderer
}

// Circle
type Circle struct {
    AbstractShapeRenderer
}

func (c *Circle) Draw() {
    c.renderer.RenderCircle()
}

// Square
type Square struct {
    AbstractShapeRenderer
}

func (s *Square) Draw() {
    s.renderer.RenderSquare()
}

// VectorRenderer
type VectorRenderer struct{}

func (vr *VectorRenderer) RenderCircle() {
    fmt.Println("Rendering Circle in Vector format")
}

func (vr *VectorRenderer) RenderSquare() {
    fmt.Println("Rendering Square in Vector format")
}

// RasterRenderer
type RasterRenderer struct{}

func (rr *RasterRenderer) RenderCircle() {
    fmt.Println("Rendering Circle in Raster format")
}

func (rr *RasterRenderer) RenderSquare() {
    fmt.Println("Rendering Square in Raster format")
}

//Drawing
type Drawing struct {
    shapes []Shape
}

func (d *Drawing) AddShape(shape Shape) {
    d.shapes = append(d.shapes, shape)
}

func (d *Drawing) DrawShapes() {
    for _, shape := range d.shapes {
        shape.Draw()
    }
}

func main() {
    drawing := Drawing{}

    circle := &Circle{}
    square := &Square{}

    vectorRenderer := &VectorRenderer{}
    rasterRenderer := &RasterRenderer{}

    circle.SetRenderer(vectorRenderer)
    square.SetRenderer(rasterRenderer)

    drawing.AddShape(circle)
    drawing.AddShape(square)

    drawing.DrawShapes()
}

File: python/bridge.py

from abc import ABC, abstractmethod

class Shape(ABC):
    @abstractmethod
    def draw(self):
        pass

class Renderer(ABC):
    @abstractmethod
    def render(self, shape):
        pass

class Circle(Shape):
    def draw(self):
        return "Drawing a circle"

class Square(Shape):
    def draw(self):
        return "Drawing a square"

class VectorRenderer(Renderer):
    def render(self, shape):
        return f"Rendering {shape.draw()} in vector format"

class RasterRenderer(Renderer):
    def render(self, shape):
        return f"Rendering {shape.draw()} in raster format"

class Drawing:
    def __init__(self):
        self.shapes = []
        self.renderer = None

    def add_shape(self, shape):
        self.shapes.append(shape)

    def set_renderer(self, renderer):
        self.renderer = renderer

    def draw_shapes(self):
        if self.renderer is None:
            raise Exception("Renderer not set")
        for shape in self.shapes:
            print(self.renderer.render(shape))

def main():
    drawing = Drawing()
    drawing.add_shape(Circle())
    drawing.add_shape(Square())

    vector_renderer = VectorRenderer()
    drawing.set_renderer(vector_renderer)
    drawing.draw_shapes()

    raster_renderer = RasterRenderer()
    drawing.set_renderer(raster_renderer)
    drawing.draw_shapes()

if __name__ == "__main__":
    main()
